### PR TITLE
CORE-548 Change ProtcolDispatcher to become a typeclass and CORE-537

### PR DESCRIPTION
**Pull request by rabbitonweb**
## Overview
Change ProtcolDispatcher to become a typeclass. This:
1. improvements code quality
2. gives us ability to abstract UnicastNetwork.receiver to a `TransportLayer`

### Does this PR relate to an RChain JIRA issue? 
CORE-548, CORE-537

### CODE REVIEW

File: comm/src/main/scala/coop/rchain/comm/network.scala [hunk 1]
```diff
@@ -42,18 +40,13 @@ class UnicastNetwork(peer: PeerNode, next: Option[ProtocolDispatcher[SocketAddre
   val comm  = new UnicastComm(local)
   val table = PeerTable(local)
 
-  def receiver[
-      F[_]: Monad: Capture: Log: Time: Metrics: TransportLayer: NodeDiscovery: Encryption: Kvs[
-        ?[_],
-        PeerNode,
-        Array[Byte]]: ApplicativeError_[?[_], CommError]: PacketHandler]: F[Unit] =
+  def receiver[F[_]: Monad: Capture: Log: Time: Metrics](
+      implicit protocolDispatcher: ProtocolDispatcher[F, SocketAddress]): F[Unit] =
```
Comment (sebtomba) [inline comm/src/main/scala/coop/rchain/comm/network.scala]: Maybe you can create a type alias to get rid of the `implicit protocolDispatcher: ProtocolDispatcher[F, SocketAddress]`?

File: comm/src/main/scala/coop/rchain/p2p/p2p.scala [hunk 2]
```diff
@@ -109,7 +103,7 @@ object Network extends ProtocolDispatcher[java.net.SocketAddress] {
 
     for {
       bootstrapAddr <- errorHandler[F].fromEither(NetworkAddress.parse(bootstrapAddrStr))
-      _             <- Log[F].info(s"Bootstrapping from $bootstrapAddr.")
+      _             <- Log[F].info(s"Bootstrapping from ${bootstrapAddr}.")
```
Comment (sebtomba) [inline comm/src/main/scala/coop/rchain/p2p/p2p.scala]: These parentheses are obsolete

File: node/src/main/scala/coop/rchain/node/node.scala [hunk 3]
```diff
@@ -59,19 +56,20 @@ class NodeRuntime(conf: Conf)(implicit scheduler: Scheduler) {
   implicit val timeEffect: Time[Task]                     = effects.time
   implicit val metricsEffect: Metrics[Task]               = effects.metrics
   implicit val inMemoryPeerKeysEffect: KeysStore[Task]    = effects.remoteKeysKvs(remoteKeysPath)
+  val net                                                 = new UnicastNetwork(src)
```
Comment (sebtomba) [inline node/src/main/scala/coop/rchain/node/node.scala]: Let's keep the implicits together. Please move it somewhere else.

File: node/src/main/scala/coop/rchain/node/node.scala [hunk 4]
```diff
@@ -59,19 +56,20 @@ class NodeRuntime(conf: Conf)(implicit scheduler: Scheduler) {
...
     com.google.protobuf.ByteString.copyFrom(Array((scala.util.Random.nextInt(10) + 1).toByte)),
     storagePath,
     storageSize
   )
-
   implicit val packetHandlerEffect: PacketHandler[Effect] = effects.packetHandler[Effect](
     casperPacketHandler[Effect]
   )
+  implicit val protocolDispatcher: ProtocolDispatcher[Effect, SocketAddress] =
+    effects.RChainProtocolDispatcher.get[Effect]
```
Comment (sebtomba) [inline node/src/main/scala/coop/rchain/node/node.scala]: Maybe you can move these implicits to the top to the others?

File: comm/src/main/scala/coop/rchain/comm/network.scala (hunk 1 above)
Comment (rabbitonweb) [reply comm/src/main/scala/coop/rchain/comm/network.scala]: ProtocolDispatcher will be removed in next PR :) [as you already know :) ]

File: comm/src/main/scala/coop/rchain/p2p/p2p.scala (hunk 2 above)
Comment (rabbitonweb) [reply comm/src/main/scala/coop/rchain/p2p/p2p.scala]: Ok

File: node/src/main/scala/coop/rchain/node/node.scala (hunk 3 above)
Comment (rabbitonweb) [reply node/src/main/scala/coop/rchain/node/node.scala]: It's part of the implementation of both TranportLayer and NodeDiscovery, that's why it's here

File: node/src/main/scala/coop/rchain/node/node.scala (hunk 4 above)
Comment (rabbitonweb) [reply node/src/main/scala/coop/rchain/node/node.scala]: Will be removed in next PR

**Pull request merged**